jobs/import_process/worker.go

```diff
@@ -0,0 +1,196 @@
+	mlog.Debug("Worker stopping", mlog.String("worker", w.name))
+	close(w.stopChan)
+	<-w.stoppedChan
+}
+
+func (w *ImportProcessWorker) doJob(job *model.Job) {
+	if claimed, err := w.jobServer.ClaimJob(job); err != nil {
+		mlog.Warn("Worker experienced an error while trying to claim job",
+			mlog.String("worker", w.name),
+			mlog.String("job_id", job.Id),
+			mlog.String("error", err.Error()))
+		return
+	} else if !claimed {
+		return
+	}
+
+	importFileName, ok := job.Data["import_file"]
+	if !ok {
+		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.missing_file", nil, "", http.StatusBadRequest)
+		w.setJobError(job, appError)
+		return
+	}
+
+	importFilePath := filepath.Join(*w.app.Config().ImportSettings.Directory, importFileName)
+	if ok, err := w.app.FileExists(importFilePath); err != nil {
+		w.setJobError(job, err)
+		return
+	} else if !ok {
+		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.file_exists", nil, "", http.StatusBadRequest)
+		w.setJobError(job, appError)
+		return
+	}
+
+	importFileSize, appErr := w.app.FileSize(importFilePath)
+	if appErr != nil {
+		w.setJobError(job, appErr)
+		return
+	}
+
+	importFile, appErr := w.app.FileReader(importFilePath)
+	if appErr != nil {
+		w.setJobError(job, appErr)
+		return
+	}
+	defer importFile.Close()
+
+	// TODO: improve this process by eliminating the need to unzip the import
+	// file locally and instead do the whole bulk import process in memory by
+	// streaming the import file.
+
+	// create a temporary dir to extract the zipped import file.
+	dir, err := ioutil.TempDir("", "import")
+	if err != nil {
+		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.tmp_dir", nil, err.Error(), http.StatusInternalServerError)
+		w.setJobError(job, appError)
+		return
+	}
+	defer os.RemoveAll(dir)
+
+	// extract the contents of the zipped file.
+	paths, err := utils.UnzipToPath(importFile.(io.ReaderAt), importFileSize, dir)
+	if err != nil {
+		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.unzip", nil, err.Error(), http.StatusInternalServerError)
+		w.setJobError(job, appError)
+		return
+	}
+
+	// find JSONL import file.
+	var jsonFilePath string
+	for _, path := range paths {
+		if filepath.Ext(path) == ".jsonl" {
+			jsonFilePath = path
+			break
+		}
+	}
+
+	if jsonFilePath == "" {
+		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.missing_jsonl", nil, "", http.StatusBadRequest)
+		w.setJobError(job, appError)
+		return
+	}
+
+	jsonFile, err := os.Open(jsonFilePath)
+	if err != nil {
+		appError := model.NewAppError("ImportProcessWorker", "import_process.worker.do_job.open_file", nil, err.Error(), http.StatusInternalServerError)
+		w.setJobError(job, appError)
+		return
+	}
+
+	// do the actual import.
+	appErr, lineNumber := w.app.BulkImportWithPath(jsonFile, false, runtime.NumCPU(), dir)
```

Comment:
We can make the workers configurable if needed. Defaulting to available CPUs for now.

Reply:
I've found in the past that for a reasonably quick bulk import you need a lot more than that workers running. not sure why - it was purely experimental rather than based on any real rationale.

Reply:
Yes, the sweet spot could depend on several factors and would need some trial and error to be figured out. Usually I would expect it to be between 2 and 3 times the available cores but that can also be quite demanding for a live system. One reason more to make it configurable I guess.

Reply:
To be clear I'm totally happy with merging with the current value. I'd rather we started slow-and-stable rather than being too aggressive and causing DB issues.